File: 239-sliding-window-maximum/239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        heap, ans = [], []
        
        for i in range(k):
            heappush(heap, (-nums[i], i))
            
        ans.append(-heap[0][0])
        
        for i in range(k, len(nums)):
            while heap and i - heap[0][1] >= k:
                heappop(heap)
                
            heappush(heap, (-nums[i], i))
            ans.append(-heap[0][0])
            
        return ans

File: 239-sliding-window-maximum/239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        multiset<int> ms;
        vector<int> res;
        
        for (int i = 0; i < k; i++)
            ms.insert(nums[i]);
        
        for (int i = k; i < (int) nums.size(); i++) {
            res.push_back(*ms.rbegin());
            ms.erase(ms.find(nums[i - k]));
            ms.insert(nums[i]);
        }
        res.push_back(*ms.rbegin());
        return res;
    }
};